// components/Carousel/Carousel.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Slider from 'react-slick';
import Image from 'components/Image';
import cx from 'classnames';
import './Carousel.css';
import 'slick-carousel/slick/slick.css';
import 'slick-carousel/slick/slick-theme.css';

export default class Carousel extends Component {
  render() {
    const { items, placeholder, className, ...settings } = this.props;

    if (items.length === 0) {
      return <Image src={placeholder} className={className} />
    }

    return (
      <div className={cx("Carousel", className)}>
        <Slider {...settings} className="Carousel-slider">
          {
            items.map((src, idx)=>(
              <div className="Carousel-item" key={idx}>
                <Image src={src} />
              </div>
            ))
          }
        </Slider>
      </div>
    );
  }
}

Carousel.propTypes = {
  className: PropTypes.string,
  items: PropTypes.array,
  speed: PropTypes.number,
  dots: PropTypes.bool,
  fade: PropTypes.bool,
  arrows: PropTypes.bool,
  infinite: PropTypes.bool,
  autoplay: PropTypes.bool,
  centerMode: PropTypes.bool,
  slidesToShow: PropTypes.number,
  slidesToScroll: PropTypes.number,
};
Carousel.defaultProps = {
  items: [],
  speed: 500,
  dots: false,
  fade: false,
  arrows: false,
  infinite: false,
  autoplay: false,
  centerMode: false,
  slidesToShow: 1,
  slidesToScroll: 1,
};
