// components/Range/Range.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import './Range.css';

const clamp = (val, min, max) => Math.min(Math.max(min, val), max);

export default class Range extends Component {
  handlePlus = () => {
    const { min, max, value, disabled } = this.props;
    !disabled && this.props.onChange(clamp(value + 1, min, max))
  };

  handleMinus = () => {
    const { min, max, value, disabled } = this.props;
    !disabled && this.props.onChange(clamp(value - 1, min, max))
  };

  render() {
    const { min, max, value, disabled } = this.props;

    return (
      <div className={cx('Range', {
        'disabled': disabled,
      })}>
        <div className={cx('Range__minus', {
          'disabled': disabled,
        })} onClick={this.handleMinus}>-</div>
        <div className="Range__input">{clamp(value, min, max)}</div>
        <div className={cx('Range__plus', {
          'disabled': disabled,
        })} onClick={this.handlePlus}>+</div>
      </div>
    );
  }
}

Range.propTypes = {
  max: PropTypes.number,
  min: PropTypes.number,
  value: PropTypes.number,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
};
Range.defaultProps = {
  value:0,
  min: 1,
  max: 100,
  disabled: false,
  onChange: ()=>{},
};
