// components/DiscountTextField/DiscountTextField.js
import React, {Component} from 'react';
import TextField from 'components/TextField';
import localizeDiscount, {
  kDiscountTypeFixedAmount,
  kDiscountTypePercentage,
} from 'services/localizeDiscount';
import './DiscountTextField.css';

//type
//amount

const discountToString = (discount) => {
  return localizeDiscount(discount.type, discount.amount);
};

const stringToDiscount = (str) => {

  if (str.match(/(\d+)([%]?)/)) {
    switch (RegExp.$2) {
      case '%':
        return {type: kDiscountTypePercentage, amount: parseInt(RegExp.$1, 10)};
      case '':
        return {
          type: kDiscountTypeFixedAmount,
          amount: parseInt(RegExp.$1, 10),
        };
    }
  }

  return null;
};

export default class DiscountTextField extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: discountToString(props.value),
    };
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      value: discountToString(nextProps.value),
    });
  }

  handleChange(e) {
    const {onChange} = this.props;
    const str = e.target.value;
    const discount = stringToDiscount(str);
    if (discount) {
      onChange(discount);
    }
  }

  render() {
    const {value} = this.state;
    return (
        <TextField {...this.props} onChange={this.handleChange} value={value}/>
    );
  }
}

DiscountTextField.propTypes = {};
DiscountTextField.defaultProps = {};
