// components/Table/Table.js
import React from 'react';
import PropTypes from 'prop-types';
import { QueryComponent, PaginationHistoryView } from 'modules/QueryController';
import Loading from 'components/Loading';
import Product from 'components/Product';
import history from 'core/history';
import cx from 'classnames';
import './Table.css';

export default class Table extends QueryComponent {
  render() {
    const { query } = this.state;
    const { className, route } = this.props;

    return (
      <div className={cx('Table-Products', className)}>
        <div className="Table-Products__grid">
          {
            query.isLoading() && <Loading size="big" color="black" />
          }
          {
            query.map( i => <Product
              key={i.id}
              className="Table-Products__grid-item"
              name={i.name}
              price={i.price}
              discount={i.discount}
              pictures={i.pictures}
              producer={i.producer}
              quantity={i.quantity}
              onClick={()=>(history.push(`/product/${i.SKU}`))}
            />)
          }
        </div>
        <PaginationHistoryView
          query={query}
          className="Table-Products__pagination"
          createLink={(page)=> `${route}/page/${page}`}
        />
      </div>
    );
  }
}

Table.propTypes = {
  page: PropTypes.number,
  query: PropTypes.func,
  objectsPerPage: PropTypes.number,
  route: PropTypes.string,
};
Table.defaultProps = {
  page: 0,
  objectsPerPage: 12,
};
