// components/Dialog/Dialog.js
import React, {Component} from 'react';
import cx from 'classnames';
import {noop} from 'underscore';
import PropTypes from 'prop-types';
import {Modal} from 'react-overlays';
import './Dialog.css';

const sizes = ['', 'lg', 'large', 'sm', 'small'];
const backdrops = ['static', true, false];

export default class Dialog extends Component {
  getChildContext() {
    return {
      hideDialog: this.props.onHide,
    };
  }

  render() {
    const {show, onHide, backdrop, closeButton, header, showHeader} = this.props;
    const {dialogClassName, headerClassName} = this.props;
    const props = {show, onHide, backdrop};

    return (
        <Modal
            aria-labelledby='modal-label'
            backdropClassName="Modal__backdrop"
            containerClassName="Modal"
            onBackdropClick={this.props.onHide}
            {...props}
        >
          <div className={cx('Modal__container', {
            'Modal__container--small': this.props.size === 'small',
            'Modal__container--large': this.props.size === 'large',
            'Modal__container--lg': this.props.size === 'lg',
            'Modal__container--sm': this.props.size === 'sm',
          })}>
            {
              showHeader &&
              <h4 className={cx('Modal__header', headerClassName)}
                  id='modal-label'>
                {header}

                {closeButton &&
                <div className='Modal__close_button' onClick={this.props.onHide}>x</div>}
              </h4>
            }
            <div className={cx('Modal__body', dialogClassName)}>
              {this.props.children}
            </div>
          </div>
        </Modal>
    );
  }
}

Dialog.propTypes = {
  header: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  onHide: PropTypes.func,
  show: PropTypes.bool,
  closeButton: PropTypes.bool,
  showHeader: PropTypes.bool,
  headerClassName: PropTypes.string,
  size: PropTypes.oneOf(sizes),
  backdrop: PropTypes.oneOf(backdrops),
};
Dialog.defaultProps = {
  size: '',
  header: '',
  show: false,
  onHide: noop,
  backdrop: 'static',
  showHeader: false,
  closeButton: true,
};
Dialog.childContextTypes = {
  hideDialog: PropTypes.func,
};
