// components/Link/Link.js
import React from 'react';
import PropTypes from 'prop-types';
import OpenNewTab from 'open-new-tab';
import history from 'core/history';

function isLeftClickEvent(event) {
  return event.button === 0;
}

function isModifiedEvent(event) {
  return !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
}

class Link extends React.Component {
  handleClick = (event) => {
    if (this.props.onClick) {
      this.props.onClick(event);
    }

    if (isModifiedEvent(event) || !isLeftClickEvent(event)) {
      return;
    }

    if (event.defaultPrevented === true) {
      return;
    }

    event.preventDefault();

    if (this.props.external) {
      OpenNewTab(this.props.to);
    } else {
      history.push(this.props.to);
    }
  };

  render() {
    const { to, children, external, ...props } = this.props;
    return <a href={to} {...props} onClick={this.handleClick}>{children}</a>;
  }
}

Link.propTypes = {
  onClick: PropTypes.func,
  external: PropTypes.bool,
  to: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};

Link.defaultProps = {
  onClick: null,
  external: false,
};



export default Link;
