// components/ReadMore/ReadMore.js
import React, {Component} from 'react';
import PropTypes from 'prop-types';

export default class ReadMore extends Component {
  constructor(props) {
    super(props);

    this.state = {
      expanded: false,
    };
  }

  toggleLines = (event) => {
    event.preventDefault();

    this.setState({
      expanded: !this.state.expanded,
    });
  };

  render() {
    const {short, full, more, less} = this.props;
    const {expanded, showLess} = this.state;

    return (
        <div>
          {!expanded && <span>{short}</span>}
          {!expanded && <span>...<a onClick={this.toggleLines}>{more}</a></span>}
          {expanded && <span>{full}</span>}
          {expanded && showLess && (
              <span><a onClick={this.toggleLines}>{less}</a></span>
          )}
        </div>
    );
  }
}

ReadMore.defaultProps = {
  short: <span/>,
  full: <span/>,
  more: 'Показать',
  less: 'Скрыть',
};

ReadMore.propTypes = {
  short: PropTypes.element,
  full: PropTypes.element,
  lines: PropTypes.number,
  showLess: PropTypes.bool,
};
