// components/ReadMore/ReadMore.js
import React, {Component} from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import './ReadMore.css';

export default class ReadMore extends Component {
  constructor(props) {
    super(props);

    this.state = {
      expanded: false,
    };
  }

  toggleLines = (e) => {
    e.preventDefault();
    e.preventDefault();

    this.setState({
      expanded: !this.state.expanded,
    });
  };

  render() {
    const {short, full, more, less} = this.props;
    const {expanded, showLess} = this.state;

    return (
        <div className={cx('ReadMore', {
          'ReadMore--expanded': expanded,
        })}>
          {!expanded && <div key="short">{short}</div>}
          {!expanded && <div key="more">...<a onClick={this.toggleLines}>{more}</a></div>}
          {expanded && <div key="full">{full}</div>}
          {expanded && showLess && <div key="less"><a onClick={this.toggleLines}>{less}</a></div>}
        </div>
    );
  }
}

ReadMore.defaultProps = {
  short: <span/>,
  full: <span/>,
  more: 'Показать полностью',
  less: 'Скрыть',
};

ReadMore.propTypes = {
  short: PropTypes.element,
  full: PropTypes.element,
  lines: PropTypes.number,
  showLess: PropTypes.bool,
};
