// components/ComboBox/ComboBox.js
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { uniqueId } from 'underscore';
import cx from 'classnames';
import './ComboBox.css';

export default class ComboBox extends PureComponent {
  render() {
    const { label, className, choices, onChange, value } = this.props;
    const name = uniqueId('comboBox');

    return (
      <div className={cx('ComboBox', className)}>
        <label htmlFor={name}>{label}</label>
        <select
          id={name}
          name={name}
          value={value}
          onChange={onChange}
          className="form-control"
        >
          {
            choices.map((i,idx) => (<option value={i.value}>{i.label}</option>))
          }
        </select>
      </div>
    );
  }
}

ComboBox.propTypes = {
  className: PropTypes.string,
  label: PropTypes.string,
  choices: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    value: PropTypes.any,
  })),
  value: PropTypes.any,
  onChange: PropTypes.func,
  allowMultiple: PropTypes.bool,
};
ComboBox.defaultProps = {

};
