// components/Address/Address.js
import React from 'react';
import Email from 'components/Email';
import Phone from 'components/Phone';
import './Address.css'

/**
 * http://microformats.org/wiki/hcard-ru
 * @param line1
 * @param line2
 * @param zip
 * @param phone
 * @param name
 * @param email
 * @returns {*}
 */
export default ({ line1, line2 = '', zip, phone, name, email, comment }) => (
    <div className="vcard">
      <strong className="n">{name}</strong>
      <div className="adr">
        {`Адрес: ${zip}, ${line1} ${line2}`}
      </div>
      <div className="tel">
        Телефон: <Phone style={{color: 'black', fontWeight: 500 }} tel={phone}/>
      </div>
      <div className="email">
        E-mail: <Email style={{color: 'black', fontWeight: 500 }} email={email}/>
      </div>

      <br/>
      {comment}
    </div>
)

/**
 *
 * @param address
 * @returns {*}
 */
export const parse = (address) => {
  if (address.manualInput) {
    return {
      line1: [address.country, address.region, address.city].join(','),
      line2: [address.street, address.house, address.room].join(','),
      name: address.name,
      phone: address.phone,
      email: address.email,
      comment: address.comment,
      zip: address.postalCode,
    }
  }

  return {
    name: address.name,
    email: address.email,
    phone: address.phone,
    line1: address.address,
    zip: address.postalCode,
    comment: address.comment,
  }
};
