// components/Sheet/Sheet.js
import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import BottomSheet from './BottomSheet';
import cx from 'classnames';
import './Sheet.css';
import {noop} from 'underscore';

const backdrops = ['static', true, false];

export default class Sheet extends PureComponent {
  getChildContext() {
    return {
      hideDialog: this.props.onHide,
    };
  }

  render() {
    const {show, onHide, backdrop, header, showHeader, headerClassName, closeButton} = this.props;
    return (
        <BottomSheet open={show}
                     onRequestClose={backdrop ? onHide : noop}
                     className="Sheet__body"
                     bottomSheetHeader={
                       showHeader ? (
                               <h4 className={cx('Sheet__header', headerClassName)}>
                                 {header}
                                 {closeButton && (
                                     <div className='Sheet__close_button'
                                          onClick={onHide}>x</div>
                                 )}
                               </h4>)
                           : null}>
          {this.props.children}
        </BottomSheet>
    );
  }
}

Sheet.propTypes = {
  header: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  onHide: PropTypes.func,
  show: PropTypes.bool,
  closeButton: PropTypes.bool,
  showHeader: PropTypes.bool,
  headerClassName: PropTypes.string,
  backdrop: PropTypes.oneOf(backdrops),
};
Sheet.defaultProps = {
  size: '',
  header: '',
  show: false,
  onHide: noop,
  backdrop: false,
  showHeader: false,
  closeButton: true,
};

Sheet.childContextTypes = {
  hideDialog: PropTypes.func,
};
