// services/localizedDelivery.js
export const kDeliveryProviderRussiaPost = "russiapost";
export const kDeliveryProviderBoxberry = "boxberry";
export const kDeliveryProviderBaikal = "baikal";
export const kDeliveryProviderPEC = "pec";
export const kDeliveryProviderNRG = "nrg";
export const kDeliveryMethodEMC = "emc";
export const kDeliveryMethodRapid = "rapid";
export const kDeliveryMethodStandard = "standard";

export default ({ provider, method }) => {
  const buffer = [];
  switch (provider) {
    case kDeliveryProviderRussiaPost:
      buffer.push("Почта России");
      switch (method) {
        case kDeliveryMethodStandard:
          buffer.push(" - Стандарт");
          break;
        case kDeliveryMethodEMC:
          buffer.push(" - Курьерская служба");
          break;
        case kDeliveryMethodRapid:
          buffer.push(" - Ускоренная");
          break;
        default: break;
      }
      break;
    case kDeliveryProviderBoxberry:
      buffer.push("Boxberry");
      break;
    case kDeliveryProviderNRG:
      buffer.push("Почтовая служба \"Энергия\"");
      break;
    case kDeliveryProviderPEC:
      buffer.push("Почтовая служба \"ПЭК\"");
      break;
    case kDeliveryProviderBaikal:
      buffer.push("Почтовая служба \"Байкал\"");
      break;
    default: break;
  }

  return buffer.join("")
}
