// services/localizedDelivery.js
export const kDeliveryProviderRussiaPost = "russiapost";
export const kDeliveryProviderBoxberry = "boxberry";
export const kDeliveryProviderBaikal = "baikal";
export const kDeliveryProviderCDEK = "cdek";
export const kDeliveryProviderPEC = "pec";
export const kDeliveryProviderNRG = "nrg";
export const kDeliveryMethodRussiaPostEMC = "pochta_emc";
export const kDeliveryMethodRussiaPostRapid = "pochta_rapid";
export const kDeliveryMethodRussiaPostStandard = "pochta_standard";
export const kDeliveryMethodCDEKEMC = "cdek_emc";
export const kDeliveryMethodCDEKRapid = "cdek_rapid";
export const kDeliveryMethodCDEKStandard = "cdek_standard";

export default ({ provider, method }) => {
  const buffer = [];
  switch (provider) {
    case kDeliveryProviderRussiaPost:
      buffer.push("Почта России");
      switch (method) {
        case kDeliveryMethodRussiaPostStandard:
          buffer.push(" - Стандарт");
          break;
        case kDeliveryMethodRussiaPostEMC:
          buffer.push(" - Курьерская служба");
          break;
        case kDeliveryMethodRussiaPostRapid:
          buffer.push(" - Ускоренная");
          break;
        default: break;
      }
      break;
    case kDeliveryProviderCDEK:
      buffer.push("CDEK");
      switch (method) {
        case kDeliveryMethodCDEKStandard:
          buffer.push(" - Стандарт");
          break;
        case kDeliveryMethodCDEKEMC:
          buffer.push(" - Курьерская служба");
          break;
        case kDeliveryMethodCDEKRapid:
          buffer.push(" - Ускоренная");
          break;
        default: break;
      }
      break;
    case kDeliveryProviderBoxberry:
      buffer.push("Boxberry");
      break;
    case kDeliveryProviderNRG:
      buffer.push("Почтовая служба \"Энергия\"");
      break;
    case kDeliveryProviderPEC:
      buffer.push("Почтовая служба \"ПЭК\"");
      break;
    case kDeliveryProviderBaikal:
      buffer.push("Почтовая служба \"Байкал\"");
      break;
    default: break;
  }

  return buffer.join("")
}
