// modules/Form/Form.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { isObject, isFunction, noop } from 'underscore';
import cx from 'classnames';

export default class Form extends Component {
  onKeyDown = (e) => {
    if (e.keyCode === 13 && this.props.enterSubmit) {
      if (e.target.tagName.toLowerCase() !== 'textarea') {
        this.onSubmit(e);
      }
    } else if (e.keyCode === 9 && this.props.tabSubmit) {
      this.props.onSubmit();
    }

    this.props.onKeyDown(e);
  };

  onSubmit = (e) => {
    if (isObject(e) && isFunction(e.preventDefault)) {
      e.preventDefault();
    }

    this.props.onSubmit();
  };

  render() {
    return (<form
      style={this.props.style}
      className={cx(this.props.className, {
        'form-horizontal': this.props.horizontal,
        'form-inline': this.props.inline,
      })}
      onKeyDown={this.onKeyDown}
      onSubmit={this.onSubmit}
    >{this.props.children}</form>);
  }

}

Form.propTypes = {
  inline: PropTypes.bool,
  horizontal: PropTypes.bool,
  style: PropTypes.object,
  tabSubmit: PropTypes.bool,
  enterSubmit: PropTypes.bool,
  onSubmit: PropTypes.func,
  onKeyDown: PropTypes.func,
};

Form.defaultProps = {
  style: {},
  horizontal: false,
  inline: false,
  tabSubmit: false,
  enterSubmit: false,
  onSubmit: noop,
  onKeyDown: noop,
};
