// dialogs/OrderEdit/constants.js
import localize, {
  kOrderStatusAwaitingFulfillment,
  kOrderStatusAwaitingPayment,
  kOrderStatusAwaitingShipment,
  kOrderStatusClosed,
  kOrderStatusCompleted,
  kOrderStatusDeclined,
  kOrderStatusRefunded,
  kOrderStatusShipped,
} from 'services/localizedOrderStatus';

export const kStatusKey = 'status';
export const kAddressKey = 'address';
export const kDiscountKey = 'discount';
export const kShippingKey = 'shipping';

export const kTotalKey = 'total';
export const kDeliveryKey = 'delivery';
export const kSubtotalKey = 'subtotal';
export const kPositionsKey = 'positions';
export const kProductPriceKey = 'productPrice';
export const kDeliveryPriceKey = 'deliveryPrice';

export const statuses = [
  kOrderStatusShipped,
  kOrderStatusRefunded,
  kOrderStatusCompleted,
  kOrderStatusDeclined,
  kOrderStatusClosed,
  kOrderStatusAwaitingShipment,
  kOrderStatusAwaitingPayment,
  kOrderStatusAwaitingFulfillment,
].map(i => (
    {value: i, label: localize(i), disabled: false, selected: false}
));
