// flux/Admin.js
import { Store } from 'modules/Flux';
import Actions from './AdminActions';
import XHR, { Methods } from './XHR';

export class Admin extends Store {
  constructor() {
    super();

    this.id = 'admin';
    this.listenables = Actions;
  }

  /**
   * Обновление заказа
   * @param order
   * @param body
   * @returns {Promise<void>}
   */
  async onUpdateOrder(order, body) {
    const { responseData } = await XHR(Methods.POST, {
      path: `/order/${order.id}`, body
    });
    this.setState({ updatedOrder: responseData })
  }

  /**
   * Обновление каталога
   * @returns {Promise<void>}
   */
  async onUpdateCatalog() {
    const { responseData } = await XHR(Methods.GET, {
      path: '/load/catalog',
    });
    console.log(responseData);
  }

  /**
   * Обновление баннеров
   * @returns {Promise<void>}
   */
  async onUpdateAds() {
    const { responseData } = await XHR(Methods.GET, {
      path: '/load/ads',
    });
    console.log(responseData);
  }

  /**
   * Удалить просроченные заказы
   * @returns {Promise<void>}
   */
  async onClearExpiredOrders() {
    const { responseData } = await XHR(Methods.POST, {
      path: '/orders/clear',
    });
    console.log(responseData);
  }

  /**
   * Создание партии
   * @returns {Promise<void>}
   */
  async onCreateBatch() {
    const { responseData } = await XHR(Methods.POST, {
      path: '/orders/batch',
    });
    console.log(responseData);
  }
}

export default new Admin();
