// flux/Cart.js
import { Store } from 'modules/Flux';
import Actions from './CartActions';
import XHR, { Methods } from './XHR';

export class Product extends Store {
  constructor() {
    super();
    this.id = 'cart';
    this.listenables = Actions;
  }

  /**
   * Загрузка корзины из сервера
   * @returns {Promise.<void>}
   */
  async onFetch() {
    const { responseData: { positions } } = await XHR(Methods.POST, { path: 'cart' });
    this.setState(this.state, { positions });
  }

  /**
   * Загрузка детальной информации о корзины
   * @returns {Promise.<void>}
   */
  async onDetail() {
    const { responseData } = await XHR(Methods.POST, { path: 'cart/detail' });
    this.setState(responseData);
  }

  /**
   * Обновить корзину
   * @param product
   * @param amount
   * @returns {Promise.<void>}
   */
  async onUpdate(product, amount) {
    const { responseData: { positions } } = await XHR(Methods.POST, {
      path: `cart/update`,
      body: {
        amount,
        productSKU: product.SKU,
        operation: 'update'}
    });
    this.setState(this.state, { positions });
  }

  /**
   * Добавить позицию
   * @param product
   * @param amount
   * @returns {Promise.<void>}
   */
  async onInsert(product, amount) {
    const { responseData: { positions } } = await XHR(Methods.POST, {
      path: `cart/update`,
      body: {
        amount,
        productSKU: product.SKU,
        operation: 'insert'}
    });
    this.setState(this.state, { positions });
  }

  /**
   * Удалить позицию
   * @param product
   * @returns {Promise.<void>}
   */
  async onDelete(product) {
    const { responseData: { positions } } = await XHR(Methods.POST, {
      path: `cart/update`,
      body: { productSKU: product.SKU, operation: 'delete' } });
    this.setState(this.state, { positions });
  }
}

export default new Product();
