// flux/QueryManager.js
import Query from 'modules/Request/Query';

export default {
  /**
   * Список коллекции
   * @returns {Promise.<void>}
   */
  queryForCollections() {
    return Query.POST('/catalog/collections');
  },

  /**
   * Список продуктов
   * @param id
   */
  queryForProducts(id) {
    return Query.POST(`/catalog/products/${id}`);
  },

  /**
   * Список распродаж
   */
  queryForSales() {
    return Query.POST('/sales');
  },

  /**
   * Список заказов
   */
  queryForOrders(status) {
    const query = Query.POST('/orders/search');
    query.appendPayload({ status });
    return query;
  },

  /**
   * Поиск заказов
   * @param filter
   * @returns {*}
   */
  queryForOrdersWithFilter(filter) {
    const query = Query.POST('/orders/search');
    query.appendPayload(filter);
    return query;
  },

  /**
   * Брендовые товары
   * @param producer
   * @returns {*}
   */
  queryForProducer(producer) {
    const query = Query.POST('/catalog/search');
    query.appendPayload({ producer });
    return query;
  },

  /**
   * Список партий
   * @param filter
   * @returns {*}
   */
  queryForBatchesWithFilter(filter) {
    const query = Query.POST('/batches/search');
    query.appendPayload(filter);
    return query;
  },
};
