// components/Currency/Currency.js
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import symbols from './symbols';
import cx from 'classnames';
import './Currency.css';
import { Percentage, FixedAmount } from './discount';

export default class Currency extends PureComponent {
  getValueWithDiscount() {
    const { discount, value } = this.props;
    if (discount) {
      switch (discount.type) {
        case Percentage:
          return (value - (value * (discount.amount / 100))).toFixed(0);
        case FixedAmount:
          return (value - discount.amount).toFixed(0);
      }
    }

    return ''
  }

  render() {
    const { discount, value } = this.props;
    const symbol = symbols[this.props.currency];

    return (
      <div className={cx('Currency')}>
        <span className={cx({
          ['Currency-strikethrough']: !!discount,
        })}>
          {value}{symbol}
        </span>
        {
          !!discount &&
          <div>
            <span>{this.getValueWithDiscount()}{symbol}</span>
          </div>
        }

      </div>
    );
  }
}

Currency.propTypes = {
  value: PropTypes.number.isRequired,
  discount: PropTypes.shape({
    type: PropTypes.string,
    amount: PropTypes.number,
  }),
  currency: PropTypes.string,
};
Currency.defaultProps = {
  value: 0,
  currency: 'RUB',
  discount: null,
};
