// modules/QueryController/QueryComponent.js
import { Component } from 'react';
import { QueryController } from 'modules/QueryController';

export default class QueryComponent extends Component {
  constructor(props) {
    super(props);

    const { objectsPerPage } = props;

    this.state = {
      ...this.getInitialState(props),
      query: new QueryController(this, {
        ...this.getControllerProps(),
        objectsPerPage
      }),
    };
  }

  getInitialState() {
    return {};
  }

  getControllerProps() {
    return {}
  }

  async componentWillMount() {
    const { page = 0 } = this.props;
    await this.state.query.loadObjectsWithPage(page, true);
  }

  async componentWillReceiveProps(newProps) {
    const { page } = newProps;
    const { query } = this.state;

    if (query.currentPage() === page) return;

    if (page) {
      query.deferLoadWithPageObjects(page, true);
    } else {
      query.deferLoadObjects();
    }
  }

  queryObjectsWillLoad(query) {
    this.setState({ query });
  }

  queryObjectsDidLoad(query) {
    this.setState({ query });
  }

  queryForRequest() {
    return null;
  }

  reloadData() {
    this.state.query.fetchCurrentPage();
  }
}
