// modules/QueryController/QueryComponent.js
import { Component } from 'react';
import { DialogController } from 'modules/Form';
import { QueryController } from 'modules/QueryController';

export default class QueryComponent extends Component {
  constructor(props) {
    super(props);

    const { objectsPerPage } = props;

    this.state = {
      ...this.getInitialState(props),
      dialogs: new DialogController(this),
      query: new QueryController(this, {
        ...this.getControllerProps(),
        objectsPerPage
      }),
    };
  }

  getInitialState() {
    return {};
  }

  getControllerProps() {
    return {}
  }

  componentWillMount() {
    const { page = 0 } = this.props;
    this.state.query.loadObjectsWithPage(page, true);
  }

  componentDidMount() {

  }

  componentWillUnmount() {

  }

  async componentWillReceiveProps(newProps) {
    const { page } = newProps;
    const { query } = this.state;

    if (query.currentPage() === page) return;

    if (page) {
      query.deferLoadWithPageObjects(page, true);
    } else {
      query.deferLoadObjects();
    }
  }

  queryObjectsWillLoad(query) {
    this.setState({ query });
  }

  queryObjectsDidLoad(query, error) {
    if (error instanceof Error) {
      this.setState({ error });
    } else {
      this.setState({ query });
    }
  }

  queryForRequest() {
    const { query } = this.props;
    return query;
  }

  // MARK: Implements Dialog Delegate
  dialogsDidChange(dialogs) {
    this.setState({dialogs});
  }

  reloadData() {
    this.state.query.fetchCurrentPage();
  }
}
