// components/Header/Header.js
import React from 'react';
import { StoreComponent } from 'modules/Flux';
import Cart from 'flux/Cart';
import Auth from 'flux/Auth';
import Link from 'components/Link';
import cx from 'classnames';
import Image from 'components/Image';
import history from 'core/history';
import './Header.css';

export default class Header extends StoreComponent {
  static menus = [
    {
      path: '/sales',
      className: "Header__item_sales",
      ic: 'ic_sales.png'
    },
    {
      path:'/payments',
      className: "Header__item_payments",
      ic:'ic_payments.png'
    },
    {
      path:'/shipment',
      className: "Header__item_shipment",
      ic:'ic_shipping.png'
    },
    {
      path: '/cart',
      className: "Header__item_cart",
      ic: 'ic_cart.png'
    },
  ];

  getInitialStore() {
    return [ Cart, Auth ];
  }

  retrieveStoreData(store, attrs) {
    if (store.id === Auth.id) {
      return {
        isAuthenticated: store.isAuthenticated(),
      }
    }
    return {}
  }

  handleClick(path) {
    return ()=>{
      history.push(path);
    }
  };

  render() {
    const { isAuthenticated } = this.state;

    return (
      <div className={cx('Header', 'Header__menu--fix', {
        'Header__menu--admin': isAuthenticated,
      })}>
        <div className="Header__menu">
          {
            Header.menus.map((i, idx)=>
              <a
                key={idx}
                className={cx('Header__item', i.className)}
                onClick={this.handleClick(i.path)}>
              </a>)
          }
        </div>
        <Link to="/admin" className="Header__admin_panel">Админ панель</Link>
        <div className="Header__section">
          <a className="Header__brand" onClick={this.handleClick('/')}>
            <Image src="ic_brand.png"/>
          </a>
        </div>
      </div>
    );
  }
}

Header.propTypes = {};
Header.defaultProps = {};
