// components/Header/Header.js
import React from 'react';
import { StoreComponent } from 'modules/Flux';
import Cart from 'flux/Cart';
import cx from 'classnames';
import Image from 'components/Image';
import history from 'core/history';
import './Header.css';

export default class Header extends StoreComponent {
  static menus = [
    {
      path: '/sales',
      className: "Header__item_sales",
      ic: 'ic_sales.png'
    },
    {
      path:'/payments',
      className: "Header__item_payments",
      ic:'ic_payments.png'
    },
    {
      path:'/shipment',
      className: "Header__item_shipment",
      ic:'ic_shipping.png'
    },
    {
      path: '/cart',
      className: "Header__item_cart",
      ic: 'ic_cart.png'
    },
  ];

  getInitialStore() {
    return [ Cart ];
  }

  handleClick(path) {
    return ()=>{
      history.push(path);
    }
  };

  render() {
    return (
      <div className="Header Header__menu-fix">
        <div className="Header__menu">
          {
            Header.menus.map((i, idx)=>
              <a
                key={idx}
                className={cx('Header__item', i.className)}
                onClick={this.handleClick(i.path)}>
              </a>)
          }
        </div>
        <div className="Header__section">
          <a className="Header__brand" onClick={this.handleClick('/')}>
            <Image src="ic_brand.png"/>
          </a>
        </div>
      </div>
    );
  }
}

Header.propTypes = {};
Header.defaultProps = {};
