// components/Header/Header.js
import React from 'react';
import { StoreComponent } from 'modules/Flux';
import Cart from 'flux/Cart';
import Image from 'components/Image';
import history from 'core/history';
import './Header.css';

export default class Header extends StoreComponent {
  static menus = [
    {
      path: '/sales',
      ic: 'ic_sales.png'
    },
    {
      path:'/payments',
      ic:'ic_payments.png'
    },
    {
      path:'/shipment',
      ic:'ic_shipping.png'
    },
    {
      path: '/cart',
      ic: 'ic_cart.png'
    },
  ];

  getInitialStore() {
    return [ Cart ];
  }

  handleClick(path) {
    return ()=>{
      history.push(path);
    }
  };

  render() {
    return (
      <div className="Header Header-menu-fix">
        <div className="Header-menu">
          {
            Header.menus.map((i, idx)=>
              <a
                key={idx}
                className="Header-item"
                onClick={this.handleClick(i.path)}>
                <Image src={i.ic}/>
              </a>)
          }
        </div>
        <div className="Header-section">
          <a className="Header-brand" onClick={this.handleClick('/')}>
            <Image src="ic_brand.png"/>
          </a>
        </div>
      </div>
    );
  }
}

Header.propTypes = {};
Header.defaultProps = {};
